src/Microsoft.TemplateEngine.Edge/Settings/AsyncMutex.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.TemplateEngine.Edge.Settings
+{
+    /// <summary>
+    /// Helper class to work with <see cref="Mutex"/> in <c>async</c> method, since <c>await</c>
+    /// can switch to different thread and <see cref="Mutex.ReleaseMutex"/> must be called from same thread.
+    /// Hence this helper class.
+    /// </summary>
+    internal sealed class AsyncMutex : IDisposable
+    {
+        private readonly TaskCompletionSource<IDisposable> _taskCompletionSource;
+        private readonly ManualResetEvent _blockReleasingMutex = new ManualResetEvent(false);
+        private readonly string _mutexName;
+        private readonly CancellationToken _token;
+        private readonly Action _unlockCallback;
+        private volatile bool _disposed;
+
+        private AsyncMutex(string mutexName, CancellationToken token, Action unlockCallback)
+        {
+            _mutexName = mutexName;
+            _token = token;
+            _unlockCallback = unlockCallback;
+            _taskCompletionSource = new TaskCompletionSource<IDisposable>();
+            ThreadPool.QueueUserWorkItem(WaitLoop);
+        }
+
+        public static Task<IDisposable> WaitAsync(string mutexName, CancellationToken token, Action unlockCallback)
+        {
+            var mutex = new AsyncMutex(mutexName, token, unlockCallback);
+            return mutex._taskCompletionSource.Task;
+        }
+
+        private void WaitLoop(object state)
+        {
+            var mutex = new Mutex(false, _mutexName);
+            bool mutexTaken = false;
+            try
+            {
+                while (true)
+                {
+                    if (_token.IsCancellationRequested)
+                    {
+                        _taskCompletionSource.SetCanceled();
+                        return;
+                    }
+                    if (mutex.WaitOne(100))
+                    {
+                        mutexTaken = true;
+                        //Check if we were cancalled while waiting for mutex...
+                        if (_token.IsCancellationRequested)
+                        {
+                            _taskCompletionSource.SetCanceled();
+                            return;
+                        }
```

Comment:
We don't need to do this check since it only partially covers the scenario and doesn't 100% work to prevent the deadlock below. We should instead apply a solution that covers all the cases (which will make it unnecessary to have this code block here).

Reply:
I realized that I was wrong about the deadlock below. But I still think that we don't need this `if` block. We have already did the work, we can return the result instead of reverting it. The caller must know that setting the CancellationToken to cancelled state doesn't always mean that the operation will be cancelled.

Reply:
If you remove the if block, you don't call `SetCancelled()`, but you `break;` out and call `SetResult(this)`. So `InstallNewPackage` continues running.

Reply:
Isn't that more error prone? Like instead of TaskCancellledException being thrown, API user will assume mutex was acquired... I guess user could check for cancellation token right after... but whats the point of having `.SetCancelled()` then?

Reply:
Applied changes and removed `if` in question.